### PR TITLE
#780: Allow users to express record matches manually at the Service and Location level

### PR DESCRIPTION
**Pull request by kdondziak**

### CODE REVIEW

File: src/main/java/org/benetech/servicenet/repository/LocationMatchRepository.java [hunk 1]
```diff
@@ -16,7 +18,10 @@
 public interface LocationMatchRepository extends JpaRepository<LocationMatch, UUID> {
     Set<LocationMatch> findByLocationId(UUID locationId);
 
-    Optional<LocationMatch> findByLocationIdAndMatchingLocationId(UUID locationId, UUID matchingLocationId);
+    @Query("SELECT locationMatch FROM LocationMatch locationMatch "
```
Comment (jkondrat) [inline src/main/java/org/benetech/servicenet/repository/LocationMatchRepository.java]: is this any different from the jpa-generated method?

File: src/main/java/org/benetech/servicenet/service/impl/LocationMatchServiceImpl.java [hunk 2]
```diff
@@ -35,6 +69,16 @@ public LocationMatch saveOrUpdate(LocationMatch locationMatch) {
 
     @Override
     public void delete(UUID locationId, UUID matchingLocationId) {
-        locationMatchRepository.deleteByLocationIdAndMatchingLocationId(locationId, matchingLocationId);
+        Optional<LocationMatch> existingMatch = locationMatchRepository.findByLocationIdAndMatchingLocationId(
```
Comment (jkondrat) [inline src/main/java/org/benetech/servicenet/service/impl/LocationMatchServiceImpl.java]: Perhaps we should add some constraint to ensure that there can be only one location/service match for given ids

Comment (jkondrat) [reply src/main/java/org/benetech/servicenet/service/impl/LocationMatchServiceImpl.java]: what would happen if we tried saving the same match concurrently?

File: src/main/webapp/app/modules/conflicts/shared/components/service/single-service-details.tsx [hunk 3]
```diff
@@ -84,14 +127,43 @@ export class SingleServiceDetails extends React.Component<ISingleServiceDetailsP
...
+    const unmatchButton = (
+      <Button onClick={this.unmatchService} replace color="danger">
+        <FontAwesomeIcon icon="arrow-left" />{' '}
+        <span className="d-none d-md-inline">
+          <Translate contentKey="multiRecordView.unmatchThisRecord" />
+        </span>
+      </Button>
+    );
+
+    const matchOrUnMatchButton = (
```
Comment (jkondrat) [inline src/main/webapp/app/modules/conflicts/shared/components/service/single-service-details.tsx]: rename to 'toggleMatch'

File: src/main/webapp/app/modules/conflicts/shared/shared-record-view.reducer.ts [hunk 4]
```diff
@@ -75,6 +91,16 @@ export default (state: SharedRecordViewState = initialState, action): SharedReco
         dismissedMatches,
         hiddenMatches
       };
+    case ACTION_TYPES.OPEN_SERVICE:
+      return {
+        ...state,
+        openService: action.payload
```
Comment (jkondrat) [inline src/main/webapp/app/modules/conflicts/shared/shared-record-view.reducer.ts]: I would rename this to openedService / selectedService

File: src/main/webapp/app/modules/conflicts/shared/shared-record-view.reducer.ts [hunk 5]
```diff
@@ -75,6 +91,16 @@ export default (state: SharedRecordViewState = initialState, action): SharedReco
...
+      };
+    case ACTION_TYPES.OPEN_LOCATION:
+      return {
+        ...state,
+        openLocation: action.payload
```
Comment (jkondrat) [inline src/main/webapp/app/modules/conflicts/shared/shared-record-view.reducer.ts]: I would rename this to openedLocation / selectedLocation (same for method names)

File: src/main/webapp/app/modules/conflicts/single/details.tsx [hunk 6]
```diff
@@ -5,13 +5,15 @@ import { RouteComponentProps, Link } from 'react-router-dom';
 import { IActivityRecord } from 'app/shared/model/activity-record.model';
 import { IOrganizationMatch } from 'app/shared/model//organization-match.model';
 import { OrganizationDetails } from '../shared/components/organization-details';
-import { LocationsDetails } from '../shared/components/location/locations-details';
-import { ServicesDetails } from '../shared/components/service/services-details';
```
Comment (jkondrat) [inline src/main/webapp/app/modules/conflicts/single/details.tsx]: what is the point of removing these braces?

File: src/main/webapp/app/modules/conflicts/single/details.tsx [hunk 7]
```diff
@@ -33,6 +35,14 @@ export class Details extends React.Component<ISingleRecordViewProp, ISingleRecor
     isAreaOpen: false
   };
 
+  componentDidMount() {
+    Promise.all([this.props.getNotHiddenMatchesByOrg(this.props.orgId)]).then(() => {
+      if (this.props.matches.length >= 0) {
+        this.props.getPartnerRecords(this.props.orgId);
```
Comment (jkondrat) [inline src/main/webapp/app/modules/conflicts/single/details.tsx]: why are we getting all partner records on single record view?

File: src/main/webapp/i18n/en/service.json [hunk 8]
```diff
@@ -42,7 +42,16 @@
       "docs": "Docs",
       "paymentsAccepted": "Payments Accepted",
       "langs": "Langs",
-      "taxonomies": "Taxonomies"
+      "taxonomies": "Taxonomies",
+      "matches": "Matches"
+    },
+    "service_match": {
```
Comment (jkondrat) [inline src/main/webapp/i18n/en/service.json]: shouldn't we use camelCase?

File: src/main/java/org/benetech/servicenet/service/mapper/OrganizationMatchMapper.java [hunk 9]
```diff
@@ -54,21 +60,27 @@
...
         OrganizationMatchDTO dto = toDto(organizationMatch);
         Organization partner = organizationMatch.getPartnerVersion();
-        Map<UUID, Set<UUID>> locationMatches = new HashMap<>();
+        Map<UUID, List<LocationMatchDto>> locationMatches = new HashMap<>();
+        Map<UUID, List<ServiceMatchDto>> serviceMatches = new HashMap<>();
+
         for (Location location : organizationMatch.getOrganizationRecord().getLocations()) {
-            Set<UUID> matches = new HashSet<>();
-            for (LocationMatch match : locationMatchService.findAllForLocation(location.getId())) {
-                if (partner.getLocations().contains(match.getMatchingLocation())) {
```
Comment (jkondrat) [inline src/main/java/org/benetech/servicenet/service/mapper/OrganizationMatchMapper.java]: Why did we stop filtering these? It was added in #801

File: src/main/java/org/benetech/servicenet/repository/LocationMatchRepository.java (hunk 1 above)
Comment (kdondziak) [reply src/main/java/org/benetech/servicenet/repository/LocationMatchRepository.java]: It's not, but tests don't see without this.

Comment (kdondziak) [reply src/main/java/org/benetech/servicenet/repository/LocationMatchRepository.java]: It's not, but tests don't see without this.

File: src/main/webapp/app/modules/conflicts/single/details.tsx (hunk 6 above)
Comment (kdondziak) [reply src/main/webapp/app/modules/conflicts/single/details.tsx]: > what is the point of removing these brackets?

Default import is not connected to redux.

**Pull request closed without merging**